Load custom formatter by relative path only if it begins with a dot

dawn-minion opened the pull request:
This commit makes it possible to load a custom formatter that is not relative to the current working directory, such as one installed
globally, or via Yarn PNP. This change makes it so we only load by a relative path if it begins with a dot, otherwise it will load
it as a regular non-relative require.

Fixes #1412

davidjgoss commented:
Would we consider making this more explicit, maybe borrowing webpack's syntax:

```shell
$ cucumber-js --format "~@my-scope/my-formatter-package"
```

```shell
$ cucumber-js --format "lib/formatters/local-custom-formatter.js"
```

(Let me know if I've missed why this wouldn't work!)

charlierudolph commented:
@davidjgoss I don't think this is about a node module vs not (as global node modules cannot be required regularly). 

@dawn-minion I don't understand why this is needed. The current implementation should work fine with absolute paths - which is what I assume you are using for globally installed modules. I'll comment on the issue as I want more specifics about the current problem you're experiencing

charlierudolph commented:
Okay plug in play is very interesting. Let's discuss this more in the issue.

dawn-minion commented:
@charlierudolph Sorry for the delay, been busy with other things. I believe that should be it here.

charlierudolph commented:
Code changes look good. Can you please add a note to the changelog under breaking changes?

dawn-minion commented:
@charlierudolph Should be ready now - had to remove the relative path parser from world.ts as well, as it would strip the leading ./ from the tests. All tests should be passing now

merceyz commented on src/formatter/builder.ts:
```diff
@@ -107,8 +107,15 @@ const FormatterBuilder = {
+      const fullCustomFormatterPath = path.resolve(cwd, customFormatterPath)
+      CustomFormatter = require(fullCustomFormatterPath) // eslint-disable-line @typescript-eslint/no-var-requires
+    } else {
+      CustomFormatter = require(customFormatterPath) // eslint-disable-line @typescript-eslint/no-var-requires
```
This should be using `createRequire` to run the require from the cwd

davidjgoss replied:
Yeah that seems a good point, although per the docs:

> ## module.createRequire(filename)#
> Added in: v12.2.0

Given we're still supporting Node 10 for the time being (its EOL is April 2021), could use https://github.com/sindresorhus/import-cwd instead.

merceyz replied:
I'd suggest using https://yarnpkg.com/package/create-require instead as it will use the node builtins when it can, so `createRequire` on node >= 12.2 `createRequireFromPath` on node >= 10.12, then a polyfil for the rest which is much better

merceyz replied:
Should actually use it for both branches here, which simplifies the logic and has the same end result

dawn-minion replied:
Added the above, works great for both cases, thanks for pointing it out

davidjgoss commented on CHANGELOG.md:
```diff
@@ -56,6 +56,7 @@ If anything is missing from the migration guide, please submit an issue.
   * Custom formatters will need to migrate
   * `json` formatter is deprecated and will be removed in next major release. Custom formatters should migrate to use the `message` formatter, or the [standalone JSON formatter](https://github.com/cucumber/cucumber/tree/master/json-formatter) as a stopgap.
 * Remove long-deprecated `typeName` from options object for `defineParameterType` in favour of `name`
+* Custom formatters are no longer loaded relative to the current directory unless it begins with a dot (e.g. `--format=./relpath/to/formatter`)
```
Worth also adding an entry to the new features section that describes what you've made possible here re PnP and just not having to do stuff like `node_modules/some-formatter/index.js`?

dawn-minion replied:
Will add here

dawn-minion replied:
Done

davidjgoss reviewed (commented):
LGTM, a couple of suggestions.

merceyz commented on src/formatter/builder.ts:
```diff
@@ -107,8 +107,15 @@ const FormatterBuilder = {
-    const CustomFormatter = require(fullCustomFormatterPath) // eslint-disable-line @typescript-eslint/no-var-requires
+    let CustomFormatter = null
+
+    if (customFormatterPath.startsWith('.')) {
+      const fullCustomFormatterPath = path.resolve(cwd, customFormatterPath)
+      CustomFormatter = require(fullCustomFormatterPath) // eslint-disable-line @typescript-eslint/no-var-requires
+    } else {
+      CustomFormatter = require(customFormatterPath) // eslint-disable-line @typescript-eslint/no-var-requires
+    }
+
```
Adding `create-require` to the dependencies and importing it this simplifies down to

```suggestion
    CustomFormatter = createRequire(cwd)(customFormatterPath)
```

dawn-minion replied:
Switched to this

dawn-minion commented:
Sorry for the delay here, been a bit busy with a move. I've added a note about Yarn PnP support to the changelog, added create-require here as @merceyz suggested which simplifies the code. Ready for review again.

merceyz commented on src/formatter/builder.ts:
```diff
@@ -107,8 +108,10 @@ const FormatterBuilder = {
   loadCustomFormatter(customFormatterPath: string, cwd: string) {
-    const fullCustomFormatterPath = path.resolve(cwd, customFormatterPath)
-    const CustomFormatter = require(fullCustomFormatterPath) // eslint-disable-line @typescript-eslint/no-var-requires
+    let CustomFormatter = null
+
+    CustomFormatter = createRequire(cwd)(customFormatterPath)
```
```suggestion
    const CustomFormatter = createRequire(cwd)(customFormatterPath)
```

dawn-minion replied:
Changed

merceyz commented on CHANGELOG.md:
```diff
@@ -56,6 +57,7 @@ If anything is missing from the migration guide, please submit an issue.
   * Custom formatters will need to migrate
   * `json` formatter is deprecated and will be removed in next major release. Custom formatters should migrate to use the `message` formatter, or the [standalone JSON formatter](https://github.com/cucumber/cucumber/tree/master/json-formatter) as a stopgap.
 * Remove long-deprecated `typeName` from options object for `defineParameterType` in favour of `name`
+* Custom formatters are no longer loaded relative to the current directory unless it begins with a dot (e.g. `--format=./relpath/to/formatter`). 
```
NIT: They are always loaded relative to the cwd now

dawn-minion replied:
Right, the note being that it uses the regular require search path relative to cwd now, instead of always trying to load a file by path from the cwd

dawn-minion replied:
Hopefully that's a bit clearer now.

davidjgoss commented on tsconfig.json:
```diff
@@ -1,7 +1,7 @@
   "compilerOptions": {
     "esModuleInterop": true,
-    "lib": ["es2017"],
+    "lib": ["es2017", "dom"],
```
Why did we add this? AFAIK we shouldn't need to rely on any browser APIs.

dawn-minion replied:
It's a requirement of the `create-require` module, unfortunately - it pulls in the "URL" type which is only included when dom is included, as far as I understand

davidjgoss replied:
Ah okay. Node has exposed `URL` on the global object since v10 according to [the docs](https://nodejs.org/api/url.html#url_the_whatwg_url_api) so I wonder if we're missing some config to make TypeScript be happy with it?

merceyz replied:
Fixed in https://github.com/nuxt-contrib/create-require/pull/3

davidjgoss replied:
Oh actually looks like an outstanding issue, see https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34960.

I think we should try to avoid having "dom" in there if possible as it would make it easy for contributors to mistakenly try to use browser-only APIs. Maybe we could override the type declaration for `create-require` to make it:

```typescript
export default function createRequire (filename: string): NodeRequire;
```

(Which is how we're using it anyway.)

merceyz replied:
Fix has been released

dawn-minion replied:
Will push an update here in just a moment

dawn-minion replied:
Done

davidjgoss replied:
That was fast! :)

dawn-minion commented:
Can't seem to reproduce the test failure on appveyor, it's passing locally here with node v10.23.0

davidjgoss reviewed (approved):
@dawn-minion I think the CI failure is an (unrelated) unstable test per #1457.
